### ISSUE TITLE
Segmentation Violation

### ISSUE THREAD

**Issue by chinochao** [sensu/sensu-go#3546]
I have build versions 5.16.1, 5.17.0 and 5.17.1 for Docker images. However, it seems that only 5.16.1 is able to run. When I try to start the backend in version 5.17.0 and 5.17.1 I get errors. I did noticed that the cache.go code was updated last month.

https://github.com/sensu/sensu-go/commit/16f7bcaa7fc1a90e9b49abd6c893089f940ec92b#diff-687fba58bf4c3563aab58eba2848851e

Here is my error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x12e8c]

goroutine 1 [running]:
github.com/sensu/sensu-go/backend/store/cache.New(0x1359108, 0x2c74060, 0x3db4000, 0x13614d8, 0x2b88230, 0x0, 0x59a1c, 0x2c74040, 0xcfe63c)
	/var/sensu/sensu-go-5.17.1/backend/store/cache/cache.go:151 +0xf4
github.com/sensu/sensu-go/backend/eventd.New(0x1359108, 0x2c30fa0, 0x1372118, 0x39b3010, 0x135d938, 0x3d26090, 0x135f918, 0x2d6ca50, 0x2b969a0, 0x3db4000, ...)
	/var/sensu/sensu-go-5.17.1/backend/eventd/eventd.go:124 +0x254
github.com/sensu/sensu-go/backend.Initialize(0x1359108, 0x2c30f80, 0x2c46000, 0x2c30f80, 0x2b99668, 0x0)
	/var/sensu/sensu-go-5.17.1/backend/backend.go:246 +0x8d0
github.com/sensu/sensu-go/backend/cmd.StartCommand.func1(0x2c10280, 0x2ba2c00, 0x0, 0xb, 0x0, 0x0)
	/var/sensu/sensu-go-5.17.1/backend/cmd/start.go:234 +0xe50
github.com/spf13/cobra.(*Command).execute(0x2c10280, 0x2ba2ba0, 0xb, 0xc, 0x2c10280, 0x2ba2ba0)
	/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x34c
github.com/spf13/cobra.(*Command).ExecuteC(0x2c10140, 0x2bbdef0, 0x1, 0x1)
	/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x230
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
	/var/sensu/sensu-go-5.17.1/cmd/sensu-backend/main.go:28 +0x114
```

Is anyone else seeing this error?

**Comment by Deltik** [comment]
I get a similar-looking crash roughly every 6 minutes on version 5.17.1.  This is on x86-64, though:

```
Feb 07 09:22:20 monitoring sensu-backend[32594]: panic: runtime error: invalid memory address or nil pointer dereference
Feb 07 09:22:20 monitoring sensu-backend[32594]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12e508f]
Feb 07 09:22:20 monitoring sensu-backend[32594]: goroutine 172 [running]:
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/bbolt.(*Tx).Bucket(...)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/bbolt@v1.3.3/tx.go:101
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/mvcc/backend.(*readTx).UnsafeRange(0xc0002e19a0, 0x28cc090, 0x3, 0x3, 0xc00148dde0, 0x11, 0x12, 0x0, 0x0, 0x0, ...)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/mvcc/backend/read_tx.go:75 +0x66f
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/mvcc.(*storeTxnRead).rangeKeys(0xc000c7aba0, 0xc00148dd00, 0x13, 0x20, 0xc00148dd40, 0x12, 0x20, 0xa3, 0x0, 0x0, ...)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/mvcc/kvstore_txn.go:140 +0x287
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/mvcc.(*storeTxnRead).Range(0xc000c7aba0, 0xc00148dd00, 0x13, 0x20, 0xc00148dd40, 0x12, 0x20, 0x0, 0x0, 0x0, ...)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/mvcc/kvstore_txn.go:45 +0xaf
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/mvcc.(*metricsTxnWrite).Range(0xc000c7abd0, 0xc00148dd00, 0x13, 0x20, 0xc00148dd40, 0x12, 0x20, 0x0, 0x0, 0x0, ...)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/mvcc/metrics_txn.go:38 +0xaf
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/etcdserver.(*applierV3backend).Range(0xc0002f89a0, 0x1d37f80, 0xc000c7abd0, 0xc0002773b0, 0x0, 0x0, 0x0)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/etcdserver/apply.go:269 +0x13f
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/etcdserver.(*EtcdServer).Range.func3()
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/etcdserver/v3_server.go:103 +0x60
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/etcdserver.(*EtcdServer).doSerialize(0xc000230f00, 0x1d37a80, 0xc0002ba940, 0xc001163330, 0xc001163348, 0x7ff3eec28fff, 0x3)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/etcdserver/v3_server.go:542 +0xec
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/etcdserver.(*EtcdServer).Range(0xc000230f00, 0x1d37a80, 0xc0002ba940, 0xc0002773b0, 0x0, 0x0, 0x0)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/etcdserver/v3_server.go:104 +0x1be
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/etcdserver/api/v3rpc.(*kvServer).Range(0xc0002baa40, 0x1d37a80, 0xc0002ba940, 0xc0002773b0, 0x7ff3f0c3cf80, 0x0, 0x414587)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/etcdserver/api/v3rpc/key.go:52 +0x78
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/proxy/grpcproxy/adapter.(*kvs2kvc).Range(0xc000a680e0, 0x1d37a80, 0xc0002ba940, 0xc0002773b0, 0x0, 0x0, 0x0, 0x7ff3f0c3cf80, 0x413ed3, 0xc000b2db70)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/proxy/grpcproxy/adapter/kv_client_adapter.go:32 +0x51
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/clientv3.(*kv).Do(0xc000a5c780, 0x1d37a80, 0xc0002ba940, 0x1, 0xc00148dd00, 0x13, 0x20, 0xc00148dd40, 0x12, 0x20, ...)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/clientv3/kv.go:149 +0x1e4
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/coreos/etcd/clientv3.(*kv).Get(0xc000a5c780, 0x1d37a80, 0xc0002ba940, 0xc00148dcc0, 0x13, 0xc000635fa0, 0x3, 0x4, 0x1786ee0, 0xc001163a88, ...)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/coreos/etcd@v3.3.17+incompatible/clientv3/kv.go:119 +0x1c7
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/sensu/sensu-go/backend/store/etcd.List(0x1d37a80, 0xc0002ba940, 0xc0002f0000, 0xc001163ba8, 0x1559880, 0xc000635f40, 0xc001163bc0, 0x0, 0x0)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/sensu/sensu-go@v0.0.0-20200131164840-40b1d5938251/backend/store/etcd/store.go:204 +0x5f9
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/sensu/sensu-go/backend/store/cache.getResources(0x1d37a80, 0xc0002ba940, 0xc0002f0000, 0x1d4b180, 0xc000bc4000, 0x1, 0x1, 0x0, 0xc001163d60, 0x4414b4)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/sensu/sensu-go@v0.0.0-20200131164840-40b1d5938251/backend/store/cache/cache.go:117 +0x26f
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/sensu/sensu-go/backend/store/cache.(*Resource).rebuild(0xc000c98300, 0x1d37a80, 0xc0002ba940, 0x0, 0x0, 0x0)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/sensu/sensu-go@v0.0.0-20200131164840-40b1d5938251/backend/store/cache/cache.go:253 +0x103
Feb 07 09:22:20 monitoring sensu-backend[32594]: github.com/sensu/sensu-go/backend/store/cache.(*Resource).start(0xc000c98300, 0x1d37a80, 0xc0002ba940)
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/sensu/sensu-go@v0.0.0-20200131164840-40b1d5938251/backend/store/cache/cache.go:239 +0x161
Feb 07 09:22:20 monitoring sensu-backend[32594]: created by github.com/sensu/sensu-go/backend/store/cache.New
Feb 07 09:22:20 monitoring sensu-backend[32594]:         /go/pkg/mod/github.com/sensu/sensu-go@v0.0.0-20200131164840-40b1d5938251/backend/store/cache/cache.go:153 +0x158
```

I attempted to work around this by having systemd restart `sensu-backend.service` every time it crashed, but sometimes the backend just doesn't respond even though the service is running.  As the workaround wasn't acceptable, I've downgraded to version 5.16.1 for the time being.

**Comment by palourde** [comment]
@Deltik Do you mind opening a new issue for this crash? Yours is happening on a completely different codebase and a different architecture so it will help us keep track of both problems.

@chinochao So this appears to be an alignment issue on 32-bit ARM. In this exact instance, I believe a possible fix could be to move the [`count` attribute](https://github.com/sensu/sensu-go/blob/40b1d59382517863bd0ba38cdc2241d6d6c1196e/backend/store/cache/cache.go#L100) at the top of the struct. That being said, it's important to note that sensu-backend is not tested at all nor supported against this CPU architecture so it's possible there's similar bugs laying around. Also, `etcd` support (which is embedded) is considered unstable too on 32-bit ARM.

**Comment by ccressent** [comment]
@chinochao I see you've editied the issue title; can you confirm which architecture you're using?

There might be a way to confirm that it's [the `sync/atomic` bug](https://golang.org/pkg/sync/atomic/#pkg-note-BUG) @palourde is mentioning. There is a `go vet` [analysis pass called `atomicalign`](https://github.com/golang/tools/commit/24cd39ecf74570612483b772d45cf9d53d908340) that may be able to confirm that. The issue is that it has to run on an ARM 32bit architecture; one cannot just set `GOARCH` to make the analysis pass pretend like it's one of the affected targets.

Also, `atomicalign` [might not be available in all the `vet` binaries](https://github.com/golang/go/issues/31886), but that should be fixed eventually. One can run that analysis pass in a standalone fashion, [as it was attempted in this issue](https://github.com/golang/go/issues/34645).

This may be a way to confirm that it is indeed the root issue before reordering struct fields.

**Comment by echlebek** [comment]
There is a fairly good chance that this issue was resolved in 5.18.1.

**Comment by palourde** [comment]
I do believe it has been fixed in 5.18.1, if not please let us know and we can re-open this issue!
